### PR TITLE
refactor(contracts): optimize code

### PR DESCRIPTION
**Pull request by ctrlc03**
# Description

Optimize code by using immutable variables are removing redundant code.

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://github.com/privacy-scaling-explorations/maci/blob/dev/CONTRIBUTING.md) and [code of conduct](https://github.com/privacy-scaling-explorations/maci/blob/dev/CODE_OF_CONDUCT.md).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: website/versioned_docs/version-v1.x/contracts.md [hunk 1]
```diff
@@ -31,8 +31,6 @@ constructor(
     signUpGatekeeper = _signUpGatekeeper;
     initialVoiceCreditProxy = _initialVoiceCreditProxy;
 
-    signUpTimestamp = block.timestamp;
```
Comment (0xmad) [inline website/versioned_docs/version-v1.x/contracts.md:34]: Is it supposed to be removed?

**Review by 0xmad — APPROVED** [review]
@ctrlc03 thanks, just one question. Do we need to support the same state for deployed smart contracts? As far as I know, if someone tries to redeploy these contracts using proxies, it won't work because state if not the same.
In case if we don't care about it, I'm good with these changes.

File: website/versioned_docs/version-v1.x/contracts.md (hunk 1 above)
Comment (ctrlc03) [reply website/versioned_docs/version-v1.x/contracts.md:34]: yes I removed it from the code as well as it wasn't used anywhere. Let me know if you think we should have kept it in the code

**Comment by ctrlc03** [comment]
> @ctrlc03 thanks, just one question. Do we need to support the same state for deployed smart contracts? As far as I know, if someone tries to redeploy these contracts using proxies, it won't work because state if not the same. In case if we don't care about it, I'm good with these changes.

Could you please expand on that? do you mean because of immutable parameters?

**Comment by 0xmad** [comment]
> > @ctrlc03 thanks, just one question. Do we need to support the same state for deployed smart contracts? As far as I know, if someone tries to redeploy these contracts using proxies, it won't work because state if not the same. In case if we don't care about it, I'm good with these changes.
> 
> Could you please expand on that? do you mean because of immutable parameters? if so, there's libraries for proxy clones with immutable args: https://github.com/wighawag/clones-with-immutable-args

I had some issues after adding immutable which causes storage layout change. But if it's not a case, let's leave it.

**Comment by ctrlc03** [comment]
> > > @ctrlc03 thanks, just one question. Do we need to support the same state for deployed smart contracts? As far as I know, if someone tries to redeploy these contracts using proxies, it won't work because state if not the same. In case if we don't care about it, I'm good with these changes.
> > 
> > 
> > Could you please expand on that? do you mean because of immutable parameters? if so, there's libraries for proxy clones with immutable args: https://github.com/wighawag/clones-with-immutable-args
> 
> I had some issues after adding immutable which causes storage layout change. But if it's not a case, let's leave it.

issues when upgrading?

**Comment by 0xmad** [comment]
> > > > @ctrlc03 thanks, just one question. Do we need to support the same state for deployed smart contracts? As far as I know, if someone tries to redeploy these contracts using proxies, it won't work because state if not the same. In case if we don't care about it, I'm good with these changes.
> > > 
> > > 
> > > Could you please expand on that? do you mean because of immutable parameters? if so, there's libraries for proxy clones with immutable args: https://github.com/wighawag/clones-with-immutable-args
> > 
> > 
> > I had some issues after adding immutable which causes storage layout change. But if it's not a case, let's leave it.
> 
> issues when upgrading? I've not tried these contracts with any proxy pattern before tbf

Yes, it's related to upgrades

**Comment by ctrlc03** [comment]
> > > > > @ctrlc03 thanks, just one question. Do we need to support the same state for deployed smart contracts? As far as I know, if someone tries to redeploy these contracts using proxies, it won't work because state if not the same. In case if we don't care about it, I'm good with these changes.
> > > > 
> > > > 
> > > > Could you please expand on that? do you mean because of immutable parameters? if so, there's libraries for proxy clones with immutable args: https://github.com/wighawag/clones-with-immutable-args
> > > 
> > > 
> > > I had some issues after adding immutable which causes storage layout change. But if it's not a case, let's leave it.
> > 
> > 
> > issues when upgrading? I've not tried these contracts with any proxy pattern before tbf
> 
> Yes, it's related to upgrades

oh that makes sense then! I don't think MACI as a protocol should use upgradable proxies, to enhance transparency. On the other hand, at some point we should definitely consider using clone proxies (or another efficient method) to deploy instances. What do you think?

**Comment by 0xmad** [comment]
@ctrlc03 I think yes, at least we need to know community about that because someone else can use it with proxies and it can be a problem after upgrade with storage layout.
From our side, I think it will be useful to have something which helps to reduce gas for deployment or even better user can just use already deployed contracts.

**Comment by ctrlc03** [comment]
> @ctrlc03 I think yes, at least we need to know community about that because someone else can use it with proxies and it can be a problem after upgrade with storage layout. From our side, I think it will be useful to have something which helps to reduce gas for deployment or even better user can just use already deployed contracts.

Definitely agree! Yes being able to use already deployed contracts and just point a proxy to it would be fantastic. Let's merge this then?

**Comment by 0xmad** [comment]
@ctrlc03 sure, thanks!

File: contracts/contracts/MACI.sol [hunk 2]
```diff
@@ -38,31 +38,27 @@ contract MACI is IMACI, Params, Utilities, Ownable {
...
   /// @notice The number of signups
-  uint256 public override numSignUps;
+  uint256 public numSignUps;
 
   /// @notice A mapping of block timestamps to the number of state leaves
   mapping(uint256 => uint256) public numStateLeaves;
 
   // ERC20 contract that hold topup credits
-  TopupCredit public topupCredit;
+  TopupCredit public immutable topupCredit;
```
Comment (kittybest) [inline contracts/contracts/MACI.sol:47]: image

Does this matter? or just leave it?

Comment (ctrlc03) [reply contracts/contracts/MACI.sol:47]: I think we can can change at some other point

Comment (samajammin) [reply contracts/contracts/MACI.sol:48]: Should we create an issue for this @kittybest @ctrlc03 @0xmad?

Comment (samajammin) [reply contracts/contracts/MACI.sol:48]: Created on here: https://github.com/privacy-scaling-explorations/maci/issues/1012

**Pull request merged**